flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java

```diff
@@ -495,19 +502,27 @@ private DataStream<RowData> distributeDataStream(DataStream<RowData> input,
   }
 
   static RowType toFlinkRowType(Schema schema, TableSchema requestedSchema) {
-    if (requestedSchema != null) {
+    return toFlinkRowType(schema, requestedSchema, null);
+  }
+
+  static RowType toFlinkRowType(Schema schema, TableSchema tableSchema, ResolvedSchema resolvedTableSchema) {
+    if (resolvedTableSchema != null) {
       // Convert the flink schema to iceberg schema firstly, then reassign ids to match the existing iceberg schema.
-      Schema writeSchema = TypeUtil.reassignIds(FlinkSchemaUtil.convert(requestedSchema), schema);
+      Schema writeSchema = TypeUtil.reassignIds(FlinkSchemaUtil.convert(resolvedTableSchema), schema);
       TypeUtil.validateWriteSchema(schema, writeSchema, true, true);
 
       // We use this flink schema to read values from RowData. The flink's TINYINT and SMALLINT will be promoted to
       // iceberg INTEGER, that means if we use iceberg's table schema to read TINYINT (backend by 1 'byte'), we will
       // read 4 bytes rather than 1 byte, it will mess up the byte array in BinaryRowData. So here we must use flink
       // schema.
-      return (RowType) requestedSchema.toRowDataType().getLogicalType();
-    } else {
-      return FlinkSchemaUtil.convert(schema);
+      return (RowType) resolvedTableSchema.toPhysicalRowDataType().getLogicalType();
```

Comment:
is there use toSourceRowDataType better? toPhysicalRowDataType has filter(TableColumn::isPhysical) bug toRowDataType not

Reply:
We don't seem to need computed, and metadata columns. As toRowDataType comments,  toPhysicalRowDataType() might be more appropriate.